vault/ssorolecredentialsprovider.go

```diff
@@ -0,0 +1,272 @@
+	ssoClientType         = "public"
+	oAuthTokenGrantType   = "urn:ietf:params:oauth:grant-type:device_code"
+	authorizationTemplate = `
+Attempting to automatically open the SSO authorization page in your default
+browser. If the browser does not open or you wish to use a different device to
+authorize this request, open the following URL:
+
+%s
+(Use Ctrl-C to abort)
+
+`
+)
+
+// CachedSSORoleCredentialsProvider uses the keyring to cache SSO Role sessions.
+type CachedSSORoleCredentialsProvider struct {
+	CredentialsName string
+	Keyring         *CredentialKeyring
+	Provider        *SSORoleCredentialsProvider
+	ExpiryWindow    time.Duration
+	credentials.Expiry
+}
+
+// Retrieve the cached credentials or generate new ones.
+func (p *CachedSSORoleCredentialsProvider) Retrieve() (credentials.Value, error) {
+	sessions := p.Keyring.Sessions()
+
+	session, err := sessions.Retrieve(p.CredentialsName, "")
+	if err != nil {
+		// session lookup missed, we need to create a new one.
+		session, err = p.Provider.GetRoleCredentials()
+		if err != nil {
+			return credentials.Value{}, err
+		}
+
+		err = sessions.Store(p.CredentialsName, "", session)
+		if err != nil {
+			return credentials.Value{}, err
+		}
+	} else {
+		log.Printf("Re-using cached credentials %s generated from GetRoleCredentials, expires in %s", FormatKeyForDisplay(*session.AccessKeyId), time.Until(*session.Expiration).String())
+	}
+
+	p.SetExpiration(*session.Expiration, p.ExpiryWindow)
+
+	return credentials.Value{
+		AccessKeyID:     *session.AccessKeyId,
+		SecretAccessKey: *session.SecretAccessKey,
+		SessionToken:    *session.SessionToken,
+	}, nil
+}
+
+// SSORoleCredentialsProvider creates temporary credentials for an SSO Role.
+type SSORoleCredentialsProvider struct {
+	OIDCProvider *SSOOIDCProvider
+	SSOClient    *sso.SSO
+	AccountID    string
+	RoleName     string
+	ExpiryWindow time.Duration
+	credentials.Expiry
+}
+
+// Retrieve generates a new set of temporary credentials using SSO GetRoleCredentials.
+func (p *SSORoleCredentialsProvider) Retrieve() (credentials.Value, error) {
+	creds, err := p.GetRoleCredentials()
+	if err != nil {
+		return credentials.Value{}, err
+	}
+
+	p.SetExpiration(*creds.Expiration, p.ExpiryWindow)
+	return credentials.Value{
+		AccessKeyID:     *creds.AccessKeyId,
+		SecretAccessKey: *creds.SecretAccessKey,
+		SessionToken:    *creds.SessionToken,
+	}, nil
+}
+
+func (p *SSORoleCredentialsProvider) GetRoleCredentials() (*sts.Credentials, error) {
+	token, err := p.OIDCProvider.GetAccessToken()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := p.SSOClient.GetRoleCredentials(&sso.GetRoleCredentialsInput{
+		AccessToken: aws.String(token.Token),
+		AccountId:   aws.String(p.AccountID),
+		RoleName:    aws.String(p.RoleName),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	expiration := aws.MillisecondsTimeValue(resp.RoleCredentials.Expiration)
+
+	// This is needed because sessions.Store expects a sts.Credentials object.
```

Comment:
yes I'm hoping to refactor sessions.Store completely and turn it into an ordinary `Provider`